db/range_del_aggregator.cc

```diff
@@ -9,233 +9,146 @@
+        return true;
       }
-      --tombstone_map_iter;
-      break;
-  }
-  assert(mode != RangeDelPositioningMode::kFullScan);
-  assert(tombstone_map_iter != tombstone_map.end() &&
-         icmp_.user_comparator()->Compare(tombstone_map_iter->first,
-                                          parsed.user_key) <= 0);
-  assert(std::next(tombstone_map_iter) == tombstone_map.end() ||
-         icmp_.user_comparator()->Compare(
-             parsed.user_key, std::next(tombstone_map_iter)->first) < 0);
-  return parsed.sequence < tombstone_map_iter->second.seq_;
-}
-
-bool RangeDelAggregator::IsRangeOverlapped(const Slice& start,
-                                           const Slice& end) {
-  // so far only implemented for non-collapsed mode since file ingestion (only
-  //  client) doesn't use collapsing
-  assert(!collapse_deletions_);
-  if (rep_ == nullptr) {
+    }
     return false;
   }
-  for (const auto& seqnum_and_tombstone_map : rep_->stripe_map_) {
-    for (const auto& start_key_and_tombstone :
-         seqnum_and_tombstone_map.second.raw_map) {
-      const auto& tombstone = start_key_and_tombstone.second;
-      if (icmp_.user_comparator()->Compare(start, tombstone.end_key_) < 0 &&
-          icmp_.user_comparator()->Compare(tombstone.start_key_, end) <= 0 &&
-          icmp_.user_comparator()->Compare(tombstone.start_key_,
-                                           tombstone.end_key_) < 0) {
+
+  bool IsRangeOverlapped(const Slice& start, const Slice& end) {
+    for (const auto& tombstone : rep_) {
+      if (ucmp_->Compare(start, tombstone.end_key_) < 0 &&
+          ucmp_->Compare(tombstone.start_key_, end) <= 0 &&
+          ucmp_->Compare(tombstone.start_key_, tombstone.end_key_) < 0) {
         return true;
       }
     }
-  }
-  return false;
-}
-
-bool RangeDelAggregator::ShouldAddTombstones(
-    bool bottommost_level /* = false */) {
-  // TODO(andrewkr): can we just open a file and throw it away if it ends up
-  // empty after AddToBuilder()? This function doesn't take into subcompaction
-  // boundaries so isn't completely accurate.
-  if (rep_ == nullptr) {
     return false;
   }
-  auto stripe_map_iter = rep_->stripe_map_.begin();
-  assert(stripe_map_iter != rep_->stripe_map_.end());
-  if (bottommost_level) {
-    // For the bottommost level, keys covered by tombstones in the first
-    // (oldest) stripe have been compacted away, so the tombstones are obsolete.
-    ++stripe_map_iter;
-  }
-  while (stripe_map_iter != rep_->stripe_map_.end()) {
-    if (!stripe_map_iter->second.raw_map.empty()) {
-      return true;
-    }
-    ++stripe_map_iter;
+
+  void AddTombstone(RangeTombstone tombstone) { rep_.emplace(tombstone); }
+
+  size_t Size() const { return rep_.size(); }
+
+  void InvalidatePosition() {}  // no-op
+
+  typedef std::multiset<RangeTombstone, TombstoneStartKeyComparator> Rep;
+
+  class Iterator : public RangeDelIterator {
```

Comment:
Can `Iterator` and `Rep` be private?

Reply:
`Iterator` is only used as ` RangeDelIterator` by other classes. I think it should be fine to make `Iterator` private and make it a friend of the parent class (via `friend class Iterator`) so that it can access `Rep`. My old-school C++ style is to put these private declarations at the top of the parent class before the `public` (i.e. before line 26).

Reply:
You're right. This works.